### DISCUSSION TITLE
Save sensor data to file

### DISCUSSION THREAD

**Discussion by mbezuidenhout** [arendst/Tasmota#13793] [Ideas]
Is there anything in the works to save sensor data to the filesystem? I see in the tasmota_globals.h file there is a define for TASM_FILE_SENSOR.
The idea is to store sensor data to file if the MQTT server has been configured but it is unable to connect for whatever reason.
It can then do a round robin so only X number of data points are stored.
It can also decrease the resolution of the data points each time the number of data points has been taken.
Once a connection can be made with the MQTT server again then the data gets sent up over a period of time.

I would like to start working on it if there isn't another project already in the works.

**Comment by mbezuidenhout** [comment]
What could also be done is have an option of downloading the sensor data from the device as a CSV. To save space it would obviously store the data as binary on the local filesystem.

**Comment by s-hadinger** [comment]
There is no plan currently. You could probably do this with Berry scripting. 

The main problem is that writing too often to flash kills it so the best model would be to keep in ram and store at regular time (ex every hour and at each restart)

**Reply by mbezuidenhout** [reply]
I agree with you about unnecessary writes to flash.
I have several NodeMCU/ESP8266 devices that I would like to implement this on, each with different sensors connected to them.

**Comment by mbezuidenhout** [comment]
The sensor data would have to have timestamps for it to mean anything. So the clock must be set before any data points can be stored.

**Reply by joba-1** [reply]
sensor data with uptime can already be useful (and could be correlated later when you get a valid time)

**Reply by mbezuidenhout** [reply]
Agreed, if it stores those data points in RAM only is can then amend the timestamps when it gets an update from a timeserver. So unless it has a timestamp that is after, say 2020, it will not get written to the filesystem.

**Comment by tyeth** [comment]
Hey @mbezuidenhout I've got similar desires, did you get anywhere with this?

**Reply by sfromis** [reply]
You have good chances of coding this with ESP32 and Berry

**Reply by mbezuidenhout** [reply]
Unfortunately I have not been able to start work on this.
